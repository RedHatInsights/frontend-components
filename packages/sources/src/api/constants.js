export const COST_MANAGEMENT_API_BASE = '/api/cost-management/v1';
