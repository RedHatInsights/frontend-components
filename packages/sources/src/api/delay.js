export const delay = (interval, prms) => {
    return new Promise(resolve => {
        setTimeout(resolve.bind(null, prms), interval);
    });
};
