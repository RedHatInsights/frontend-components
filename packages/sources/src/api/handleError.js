import { getSourcesApi } from './index';

export const handleError = (id, error) => {
    if (!error) {
        return 'Undefined error';
    }

    if (typeof error === 'string') {
        return error;
    }

    const url = error.config.url;
    let detail;

    if (typeof error.response.data === 'string') {
        detail = error.response.data;
    } else {
        detail = error.response.data.errors[0].detail;
    }

    const detailInfo = error.detailInfo ? `(${error.detailInfo}) ` : ``;

    return getSourcesApi().deleteSource(id)
    .then(() => {
        return `${detailInfo}${url}: ${detail}`;
    })
    .catch((errorDelete) => {
        return `${detailInfo}${url}: ${detail}. The source was not removed, try remove it later: ${errorDelete.response.data.errors[0].detail}`; }
    );
};
