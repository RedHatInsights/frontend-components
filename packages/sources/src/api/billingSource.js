import { axiosInstance } from './index';
import { COST_MANAGEMENT_API_BASE } from './constants';

export const postBillingSource = (data, timeout = 0, interval = 1000) => new Promise((resolve, reject) => {
    const start = Date.now();

    const postBillingSourceInner = () => axiosInstance
    .post(`${COST_MANAGEMENT_API_BASE}/sources/billing_source/`, data)
    .then(({ data }) => resolve(data))
    .catch((error) => (Date.now() - start) >= timeout ?
        reject({ ...error, detailInfo: 'Billing source timeout expired' }) :
        setTimeout(() => postBillingSourceInner(data), interval)
    );

    return postBillingSourceInner(data);
});
