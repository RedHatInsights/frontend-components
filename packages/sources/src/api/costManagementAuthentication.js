import { axiosInstance } from './index';
import { COST_MANAGEMENT_API_BASE } from './constants';

const delay = (interval, prms) => {
    return new Promise(resolve => {
        setTimeout(resolve.bind(null, prms), interval);
    });
};

export function patchSource(data, timeout = 10000, interval = 1000) {
    return new Promise((res, rej) => {
        const start = Date.now();
        const { id, ...payload } = data;

        const checkSource = () => axiosInstance
        .get(`${COST_MANAGEMENT_API_BASE}/sources/${id}/`)
        .catch(error => {
            if ((Date.now() - start) >= timeout) {
                return Promise.reject(`Cost Management failed tracking source: ${JSON.stringify(error)}`);
            }

            return delay(interval).then(() => checkSource());
        });

        return checkSource()
        .then(() => axiosInstance.patch(`${COST_MANAGEMENT_API_BASE}/sources/${id}/`, payload))
        .then(data => res(data))
        .catch(error => rej(error));
    });
}
