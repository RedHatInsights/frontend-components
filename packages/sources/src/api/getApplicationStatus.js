import { getSourcesApi } from '.';
import { delay } from './delay';

export const checkAppAvailability = (id, timeout = 10000, interval = 1000, entity = 'getApplication', updatedTime) => new Promise((res, rej) => {
    const start = Date.now();

    const checkSource = () => getSourcesApi()[entity](id).then(data => {
        const isTimeOuted = (Date.now() - start) >= timeout;

        if (isTimeOuted) {
            if (updatedTime) {
                return ({
                    ...data,
                    availability_status: null,
                    availability_status_error: null
                });
            }

            return data;
        }

        if (data.availability_status === 'available' || data.availability_status === 'unavailable') {
            if (updatedTime && new Date(data.last_checked_at || data.updated_at) > updatedTime) {
                return data;
            }

            if (!updatedTime) {
                return data;
            }
        }

        return delay(interval).then(() => checkSource());
    });

    return checkSource()
    .then(data => res(data))
    .catch(error => rej(error));
});
