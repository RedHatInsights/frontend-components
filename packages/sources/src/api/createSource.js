import { postBillingSource } from './billingSource';
import { handleError } from './handleError';

import { getSourcesApi } from './index';

export const parseUrl = url => {
    if (!url) {
        return ({});
    }

    try {
        const u = new URL(url);
        return {
            scheme: u.protocol.replace(/:$/, ''),
            host: u.hostname,
            port: u.port,
            path: u.pathname
        };
    } catch (error) {
        console.log(error);
        return ({});
    }
};

export const urlOrHost = formData => formData.url ? parseUrl(formData.url) : formData;

export const handleErrorWrapper = (sourceId) => async(error) => await handleError(error, sourceId);

export const doCreateSource = async (formData, sourceTypes) => {
    let sourceDataOut;

    try {
        const source_type_id = sourceTypes.find((x) => x.name === formData.source_type).id;

        sourceDataOut = await getSourcesApi().createSource({ ...formData.source, source_type_id  });

        const promises = [];

        if (!formData.noEndpoint) {
            const { scheme, host, port, path } = urlOrHost(formData);

            const endPointPort = parseInt(port, 10);

            const endpointData = {
                ...formData.endpoint,
                default: true,
                source_id: sourceDataOut.id,
                scheme,
                host,
                port: isNaN(endPointPort) ? undefined : endPointPort,
                path
            };

            promises.push(getSourcesApi().createEndpoint(endpointData));
        } else {
            promises.push(Promise.resolve(undefined));
        }

        if (formData.application && formData.application.application_type_id) {
            const applicationData = {
                ...formData.application,
                source_id: sourceDataOut.id
            };

            promises.push(getSourcesApi().createApplication(applicationData));
        } else {
            promises.push(Promise.resolve(undefined));
        }

        if (formData.billing_source) {
            const billingSourceData = {
                billing_source: formData.billing_source,
                source_id: sourceDataOut.id
            };

            promises.push(postBillingSource(billingSourceData));
        }

        const [ endpointDataOut, applicationDataOut ] = await Promise.all(promises);

        if (endpointDataOut) {
            const authenticationData = {
                ...formData.authentication,
                resource_id: endpointDataOut.id,
                resource_type: 'Endpoint'
            };

            await getSourcesApi().createAuthentication(authenticationData);
        }

        return {
            ...sourceDataOut,
            endpoint: [ endpointDataOut ],
            applications: [ applicationDataOut ]
        };
    } catch (error) {
        const errorMessage = await handleError(error, sourceDataOut ? sourceDataOut.id : undefined);
        throw errorMessage;
    }
};
