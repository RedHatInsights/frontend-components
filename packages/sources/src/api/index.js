import axios from 'axios';
import { DefaultApi as SourcesDefaultApi } from '@redhat-cloud-services/sources-client';
import { Base64 } from 'js-base64';
import { postBillingSource } from './billingSource';

const calculateApiBase = b => (
    (b.endsWith('/') && `${b}v1.0`) || `${b}/sources/v1.0`
);

export const SOURCES_API_BASE = calculateApiBase(process.env.BASE_PATH || '');

const axiosInstance = axios.create(
    process.env.FAKE_IDENTITY ? {
        headers: {
            common: {
                'x-rh-identity': Base64.encode(
                    JSON.stringify(
                        {
                            identity: { account_number: process.env.FAKE_IDENTITY }
                        }
                    )
                )
            }
        }
    } : {}
);

axiosInstance.interceptors.request.use(async (config) => {
    await window.insights.chrome.auth.getUser();
    return config;
});
axiosInstance.interceptors.response.use(response => response.data || response);
axiosInstance.interceptors.response.use(null, error => { throw { ...error.response }; });

export { axiosInstance };

let apiInstance;

export const getSourcesApi = () =>
    apiInstance || (apiInstance = new SourcesDefaultApi(undefined, SOURCES_API_BASE, axiosInstance));

export const doLoadSourceTypes = () =>
    getSourcesApi().listSourceTypes().then(data => ({ sourceTypes: data.data }));

export const doLoadApplicationTypes = () =>
    getSourcesApi().listApplicationTypes().then(data => ({ applicationTypes: data.data }));

const parseUrl = url => {
    if (!url) {
        return ({});
    }

    try {
        const u = new URL(url);
        return {
            scheme: u.protocol.replace(/:$/, ''),
            host: u.hostname,
            port: u.port,
            path: u.pathname
        };
    } catch (error) {
        console.log(error);
        return ({});
    }
};

/*
     * If there's an URL in the formData, parse it and use it,
     * else use individual fields (scheme, host, port, path).
     */

const urlOrHost = formData => formData.url ? parseUrl(formData.url) : formData;

export function doCreateSource(formData, sourceTypes) {
    const source_type_id = sourceTypes.find((x) => x.name === formData.source_type).id;

    return getSourcesApi().createSource({ ...formData.source, source_type_id }).then((sourceDataOut) => {
        const { scheme, host, port, path } = urlOrHost(formData);

        const endPointPort = parseInt(port, 10);

        const endpointData = {
            ...formData.endpoint,
            default: true,
            source_id: sourceDataOut.id,
            scheme,
            host,
            port: isNaN(endPointPort) ? undefined : endPointPort,
            path
        };

        const promises = [ getSourcesApi().createEndpoint(endpointData) ];

        if (formData.application.application_type_id) {
            const applicationData = {
                ...formData.application,
                source_id: sourceDataOut.id
            };

            promises.push(getSourcesApi().createApplication(applicationData));
        }

        return Promise.all(promises).then(([ endpointDataOut, applicationDataOut = undefined ]) => {
            const authenticationData = {
                ...formData.authentication,
                resource_id: endpointDataOut.id,
                resource_type: 'Endpoint'
            };

            return getSourcesApi().createAuthentication(authenticationData).then(() => {
                const source = {
                    ...sourceDataOut,
                    endpoint: [ endpointDataOut ],
                    applications: [ applicationDataOut ]
                };

                if (formData.billing_source) {
                    const billingSourceData = {
                        billing_source: formData.billing_source,
                        source_id: sourceDataOut.id
                    };

                    return postBillingSource(billingSourceData).then(() => source)
                    .catch(error => { throw { error };});
                }

                return source;
            }, (error) => {
                console.error('Authentication creation failure:', error);
                throw { error };
            });
        }).catch(e => {
            console.error(e);
            throw { error: e };
        });
    }, (error) => {
        console.error('Source creation failure:', error);
        throw { error };
    });
}

export const doLoadAllApplications = () => getSourcesApi().listApplicationTypes().then(data => data.data);

export const findSource = (name) => getSourcesApi().postGraphQL({
    query: `{ sources(filter: {name: "${name}"})
        { id, name }
    }`
});
