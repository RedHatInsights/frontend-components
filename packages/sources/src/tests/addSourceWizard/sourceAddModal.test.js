import React from 'react';
import { shallow } from 'enzyme';
import { Wizard } from '@patternfly/react-core';

import AddSourceWizard from '../../addSourceWizard/SourceAddModal';
import sourceTypes from '../helpers/sourceTypes';
import applicationTypes from '../helpers/applicationTypes';
import SourcesFormRenderer from '../../sourceFormRenderer/index';

import * as dependency from '../../api/index';

describe('Steps components', () => {
    let initialProps;
    let spyFunction;

    beforeEach(() => {
        spyFunction = jest.fn();

        initialProps = {
            isOpen: true,
            refreshSources: spyFunction,
            onCancel: jest.fn(),
            onSubmit: jest.fn()
        };
    });

    afterEach(() => {
        spyFunction.mockReset();
    });

    it('renders correctly with sourceTypes and applicationTypes', () => {
        const wrapper = shallow(<AddSourceWizard { ...initialProps } sourceTypes={ sourceTypes } applicationTypes={ applicationTypes }/>);
        expect(wrapper.find(SourcesFormRenderer)).toHaveLength(1);
    });

    it('renders correctly without sourceTypes', (done) => {
        // mock API
        dependency.doLoadApplicationTypes = jest.fn(() => new Promise((resolve) => resolve({ applicationTypes })));
        dependency.doLoadSourceTypes = jest.fn(() => new Promise((resolve) => resolve({ sourceTypes })));

        const wrapper = shallow(<AddSourceWizard { ...initialProps } applicationTypes={ applicationTypes }/>);

        // loading state
        expect(wrapper.find(SourcesFormRenderer)).toHaveLength(0);
        expect(wrapper.find(Wizard)).toHaveLength(1);

        // async call
        setImmediate(() => {
            wrapper.update();
            expect(dependency.doLoadSourceTypes).toHaveBeenCalled();
            expect(dependency.doLoadApplicationTypes).not.toHaveBeenCalled();
            expect(wrapper.find(SourcesFormRenderer)).toHaveLength(1);
            done();
        });
    });

    it('renders correctly without applicationTypes', (done) => {
        // mock API
        dependency.doLoadSourceTypes = jest.fn(() => new Promise((resolve) => resolve({ sourceTypes })));
        dependency.doLoadApplicationTypes = jest.fn(() => new Promise((resolve) => resolve({ applicationTypes })));

        const wrapper = shallow(<AddSourceWizard { ...initialProps } sourceTypes={ sourceTypes }/>);

        // loading state
        expect(wrapper.find(SourcesFormRenderer)).toHaveLength(0);
        expect(wrapper.find(Wizard)).toHaveLength(1);

        // async call
        setImmediate(() => {
            wrapper.update();
            expect(dependency.doLoadSourceTypes).not.toHaveBeenCalled();
            expect(dependency.doLoadApplicationTypes).toHaveBeenCalled();
            expect(wrapper.find(SourcesFormRenderer)).toHaveLength(1);
            done();
        });
    });

    it('renders correctly without sourceTypes and application types', (done) => {
        // mock API
        dependency.doLoadSourceTypes = jest.fn(() => new Promise((resolve) => resolve({ sourceTypes })));
        dependency.doLoadApplicationTypes = jest.fn(() => new Promise((resolve) => resolve({ applicationTypes })));

        const wrapper = shallow(<AddSourceWizard { ...initialProps }/>);

        // loading state
        expect(wrapper.find(SourcesFormRenderer)).toHaveLength(0);
        expect(wrapper.find(Wizard)).toHaveLength(1);

        // async call
        setImmediate(() => {
            wrapper.update();
            expect(dependency.doLoadSourceTypes).toHaveBeenCalled();
            expect(dependency.doLoadApplicationTypes).toHaveBeenCalled();
            expect(wrapper.find(SourcesFormRenderer)).toHaveLength(1);
            done();
        });
    });
});
