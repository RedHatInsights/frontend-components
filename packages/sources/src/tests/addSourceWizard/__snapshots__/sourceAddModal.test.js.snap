// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps components renders correctly with sourceTypes 1`] = `
<SourcesFormRenderer
  initialValues={Object {}}
  onCancel={[MockFunction]}
  onSubmit={[Function]}
  schema={
    Object {
      "fields": Array [
        Object {
          "buttonLabels": Object {
            "submit": "Finish",
          },
          "component": "wizard",
          "description": "You are importing data into this platform",
          "fields": Array [
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      To import data for an application, you need to connect to a data source. To begin, input a name and select the type of source you want to collect data from.
                    </Text>
                    <Text
                      component="p"
                    >
                      All fields are required.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, Source_1",
                  "isRequired": true,
                  "label": "Name",
                  "name": "source_name",
                  "type": "text",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
                Object {
                  "component": "card-select",
                  "iconMapper": [Function],
                  "isRequired": true,
                  "label": "Type",
                  "name": "source_type",
                  "options": Array [
                    Object {
                      "label": "OpenShift Container Platform",
                      "value": "openshift",
                    },
                    Object {
                      "label": "Amazon Web Services",
                      "value": "amazon",
                    },
                    Object {
                      "label": "Ansible Tower",
                      "value": "ansible-tower",
                    },
                  ],
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
              ],
              "name": "step_1",
              "nextStep": Object {
                "stepMapper": Object {
                  "amazon": "amazon",
                  "ansible-tower": "ansible-tower",
                  "openshift": "openshift",
                },
                "when": "source_type",
              },
              "stepKey": 1,
              "title": "Select a source type",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      Add credentials that enable communication with this source. This source requires the login token.
                    </Text>
                    <Text
                      component="p"
                    >
                      To collect data from a Red Hat OpenShift Container Platform source,
                    </Text>
                    <TextContent>
                      <TextList
                        component="ul"
                      >
                        <TextListItem
                          component="li"
                        >
                          Log in to the Red Hat OpenShift Container Platform cluster with an account that has access to the namespace
                        </TextListItem>
                        <TextListItem
                          component="li"
                        >
                          Run the following command to obtain your login token:
                          <b>
                             # oc sa get-token -n management-infra management-admin
                          </b>
                        </TextListItem>
                        <TextListItem
                          component="li"
                        >
                          Copy the token and paste it in the following field.
                        </TextListItem>
                      </TextList>
                    </TextContent>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "textarea-field",
                  "label": "Token",
                  "name": "token",
                },
              ],
              "name": "openshift",
              "nextStep": "openshift_1",
              "stepKey": "openshift",
              "title": "Add source credentials",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      Provide OpenShift Container Platform URL and SSL certificate.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "kubernetes",
                  "name": "role",
                  "type": "hidden",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "token",
                  "name": "authtype",
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, https://myopenshiftcluster.mycompany.com",
                  "isRequired": true,
                  "label": "URL",
                  "name": "url",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
                Object {
                  "component": "checkbox",
                  "label": "Verify SSL",
                  "name": "verify_ssl",
                },
                Object {
                  "component": "textarea-field",
                  "condition": Object {
                    "is": true,
                    "when": "verify_ssl",
                  },
                  "label": <SSLFormLabel />,
                  "name": "certificate_authority",
                },
              ],
              "name": "openshift_1",
              "nextStep": "summary",
              "stepKey": "openshift_1",
              "title": "Enter OpenShift Container Platform information",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      <Title
                        headingLevel="h3"
                        size="2xl"
                      >
                        Configure account access 
                        <Popover
                          appendTo={[Function]}
                          aria-label="Help text"
                          bodyContent={
                            <React.Fragment>
                              <Text
                                component="p"
                              >
                                Red Had recommends using the Power User AWS Identity and Access Management (IAM) policy when adding an AWS account as a source. This Policy allows the user full access to API functionality and AWS services for user administration.
                                <br />
                                Create an access key in the  
                                <b>
                                  Security Credentials
                                </b>
                                  area of your AWS user account. To add your account as a source, enter the access key ID and secret access key to act as your user ID and password.
                              </Text>
                            </React.Fragment>
                          }
                          className=""
                          closeBtnAriaLabel="Close"
                          enableFlip={true}
                          footerContent={
                            <a
                              href="http://foo.bar"
                            >
                              Learn more
                            </a>
                          }
                          headerContent={null}
                          hideOnOutsideClick={true}
                          isVisible={null}
                          maxWidth="50%"
                          onHidden={[Function]}
                          onHide={[Function]}
                          onMount={[Function]}
                          onShow={[Function]}
                          onShown={[Function]}
                          position="bottom"
                          shouldClose={[Function]}
                          zIndex={9999}
                        >
                          <QuestionCircleIcon
                            color="currentColor"
                            noVerticalAlign={false}
                            size="sm"
                            title={null}
                          />
                        </Popover>
                      </Title>
                    </Text>
                    <Text
                      component="p"
                    >
                      Create an access key in your AWS user account and enter the details below.
                    </Text>
                    <Text
                      component="p"
                    >
                      For sufficient access and security, Red Hat recommends using the Power User IAM polocy for your AWS user account.
                    </Text>
                    <Text
                      component="p"
                    >
                      All fields are required.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "aws",
                  "name": "role",
                  "type": "hidden",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "access_key_secret_key",
                  "name": "authtype",
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, AKIAIOSFODNN7EXAMPLE",
                  "isRequired": true,
                  "label": "Access Key ID",
                  "name": "username",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, wJairXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
                  "isRequired": true,
                  "label": "Secret Key",
                  "name": "password",
                  "type": "password",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
              ],
              "name": "amazon",
              "nextStep": "summary",
              "stepKey": "amazon",
              "title": "Configure account access",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "text-field",
                  "name": "field",
                },
              ],
              "name": "ansible-tower",
              "nextStep": "summary",
              "stepKey": "ansible-tower",
              "title": "AnsibleTower",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Title
                      headingLevel="h3"
                      size="2xl"
                    >
                      Review source details
                    </Title>
                    <Text
                      component="p"
                    >
                      Review the information below and click Finish to configure your project. Use the Back button to make changes.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "summary",
                  "name": "summary",
                  "sourceTypes": Array [
                    Object {
                      "id": "1",
                      "name": "openshift",
                      "product_name": "OpenShift Container Platform",
                      "schema": Object {
                        "fields": Array [
                          Object {
                            "component": "text-field",
                            "name": "field",
                          },
                        ],
                        "title": "openshift",
                      },
                    },
                    Object {
                      "id": "2",
                      "name": "amazon",
                      "product_name": "Amazon Web Services",
                      "schema": Object {
                        "fields": Array [
                          Object {
                            "component": "text-field",
                            "name": "field",
                          },
                        ],
                        "title": "AWS",
                      },
                    },
                    Object {
                      "id": "3",
                      "name": "ansible-tower",
                      "product_name": "Ansible Tower",
                      "schema": Object {
                        "fields": Array [
                          Object {
                            "component": "text-field",
                            "name": "field",
                          },
                        ],
                        "title": "AnsibleTower",
                      },
                    },
                  ],
                },
              ],
              "name": "summary",
              "stepKey": "summary",
              "title": "Review source details",
            },
          ],
          "inModal": true,
          "name": "wizard",
          "title": "Add a source",
        },
      ],
    }
  }
  showFormControls={false}
/>
`;

exports[`Steps components renders correctly without sourceTypes 1`] = `
<Wizard
  ariaLabelCloseButton="Close"
  ariaLabelNav="Steps"
  backButtonText="Back"
  cancelButtonText="Cancel"
  className=""
  description="You are importing data into this platform"
  footer={null}
  hasBodyPadding={true}
  height={null}
  isCompactNav={false}
  isFullHeight={false}
  isFullWidth={false}
  isOpen={true}
  nextButtonText="Next"
  onBack={null}
  onClose={[MockFunction]}
  onGoToStep={null}
  onNext={null}
  startAtStep={1}
  steps={
    Array [
      Object {
        "component": <LoadingStep
          customText="Loading, please wait."
          onClose={[MockFunction]}
        />,
        "isFinishedStep": true,
        "name": "Loading",
      },
    ]
  }
  title="Add a source"
  width={null}
/>
`;

exports[`Steps components renders correctly without sourceTypes 2`] = `
<SourcesFormRenderer
  initialValues={Object {}}
  onCancel={[MockFunction]}
  onSubmit={[Function]}
  schema={
    Object {
      "fields": Array [
        Object {
          "buttonLabels": Object {
            "submit": "Finish",
          },
          "component": "wizard",
          "description": "You are importing data into this platform",
          "fields": Array [
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      To import data for an application, you need to connect to a data source. To begin, input a name and select the type of source you want to collect data from.
                    </Text>
                    <Text
                      component="p"
                    >
                      All fields are required.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, Source_1",
                  "isRequired": true,
                  "label": "Name",
                  "name": "source_name",
                  "type": "text",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
                Object {
                  "component": "card-select",
                  "iconMapper": [Function],
                  "isRequired": true,
                  "label": "Type",
                  "name": "source_type",
                  "options": Array [
                    Object {
                      "label": "OpenShift Container Platform",
                      "value": "openshift",
                    },
                    Object {
                      "label": "Amazon Web Services",
                      "value": "amazon",
                    },
                    Object {
                      "label": "Ansible Tower",
                      "value": "ansible-tower",
                    },
                  ],
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
              ],
              "name": "step_1",
              "nextStep": Object {
                "stepMapper": Object {
                  "amazon": "amazon",
                  "ansible-tower": "ansible-tower",
                  "openshift": "openshift",
                },
                "when": "source_type",
              },
              "stepKey": 1,
              "title": "Select a source type",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      Add credentials that enable communication with this source. This source requires the login token.
                    </Text>
                    <Text
                      component="p"
                    >
                      To collect data from a Red Hat OpenShift Container Platform source,
                    </Text>
                    <TextContent>
                      <TextList
                        component="ul"
                      >
                        <TextListItem
                          component="li"
                        >
                          Log in to the Red Hat OpenShift Container Platform cluster with an account that has access to the namespace
                        </TextListItem>
                        <TextListItem
                          component="li"
                        >
                          Run the following command to obtain your login token:
                          <b>
                             # oc sa get-token -n management-infra management-admin
                          </b>
                        </TextListItem>
                        <TextListItem
                          component="li"
                        >
                          Copy the token and paste it in the following field.
                        </TextListItem>
                      </TextList>
                    </TextContent>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "textarea-field",
                  "label": "Token",
                  "name": "token",
                },
              ],
              "name": "openshift",
              "nextStep": "openshift_1",
              "stepKey": "openshift",
              "title": "Add source credentials",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      Provide OpenShift Container Platform URL and SSL certificate.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "kubernetes",
                  "name": "role",
                  "type": "hidden",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "token",
                  "name": "authtype",
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, https://myopenshiftcluster.mycompany.com",
                  "isRequired": true,
                  "label": "URL",
                  "name": "url",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
                Object {
                  "component": "checkbox",
                  "label": "Verify SSL",
                  "name": "verify_ssl",
                },
                Object {
                  "component": "textarea-field",
                  "condition": Object {
                    "is": true,
                    "when": "verify_ssl",
                  },
                  "label": <SSLFormLabel />,
                  "name": "certificate_authority",
                },
              ],
              "name": "openshift_1",
              "nextStep": "summary",
              "stepKey": "openshift_1",
              "title": "Enter OpenShift Container Platform information",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Text
                      component="p"
                    >
                      <Title
                        headingLevel="h3"
                        size="2xl"
                      >
                        Configure account access 
                        <Popover
                          appendTo={[Function]}
                          aria-label="Help text"
                          bodyContent={
                            <React.Fragment>
                              <Text
                                component="p"
                              >
                                Red Had recommends using the Power User AWS Identity and Access Management (IAM) policy when adding an AWS account as a source. This Policy allows the user full access to API functionality and AWS services for user administration.
                                <br />
                                Create an access key in the  
                                <b>
                                  Security Credentials
                                </b>
                                  area of your AWS user account. To add your account as a source, enter the access key ID and secret access key to act as your user ID and password.
                              </Text>
                            </React.Fragment>
                          }
                          className=""
                          closeBtnAriaLabel="Close"
                          enableFlip={true}
                          footerContent={
                            <a
                              href="http://foo.bar"
                            >
                              Learn more
                            </a>
                          }
                          headerContent={null}
                          hideOnOutsideClick={true}
                          isVisible={null}
                          maxWidth="50%"
                          onHidden={[Function]}
                          onHide={[Function]}
                          onMount={[Function]}
                          onShow={[Function]}
                          onShown={[Function]}
                          position="bottom"
                          shouldClose={[Function]}
                          zIndex={9999}
                        >
                          <QuestionCircleIcon
                            color="currentColor"
                            noVerticalAlign={false}
                            size="sm"
                            title={null}
                          />
                        </Popover>
                      </Title>
                    </Text>
                    <Text
                      component="p"
                    >
                      Create an access key in your AWS user account and enter the details below.
                    </Text>
                    <Text
                      component="p"
                    >
                      For sufficient access and security, Red Hat recommends using the Power User IAM polocy for your AWS user account.
                    </Text>
                    <Text
                      component="p"
                    >
                      All fields are required.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "aws",
                  "name": "role",
                  "type": "hidden",
                },
                Object {
                  "component": "text-field",
                  "initialValue": "access_key_secret_key",
                  "name": "authtype",
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, AKIAIOSFODNN7EXAMPLE",
                  "isRequired": true,
                  "label": "Access Key ID",
                  "name": "username",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
                Object {
                  "component": "text-field",
                  "helperText": "For example, wJairXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
                  "isRequired": true,
                  "label": "Secret Key",
                  "name": "password",
                  "type": "password",
                  "validate": Array [
                    Object {
                      "type": "required-validator",
                    },
                  ],
                },
              ],
              "name": "amazon",
              "nextStep": "summary",
              "stepKey": "amazon",
              "title": "Configure account access",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "text-field",
                  "name": "field",
                },
              ],
              "name": "ansible-tower",
              "nextStep": "summary",
              "stepKey": "ansible-tower",
              "title": "AnsibleTower",
            },
            Object {
              "fields": Array [
                Object {
                  "component": "description",
                  "content": <TextContent>
                    <Title
                      headingLevel="h3"
                      size="2xl"
                    >
                      Review source details
                    </Title>
                    <Text
                      component="p"
                    >
                      Review the information below and click Finish to configure your project. Use the Back button to make changes.
                    </Text>
                  </TextContent>,
                  "name": "description-summary",
                },
                Object {
                  "component": "summary",
                  "name": "summary",
                  "sourceTypes": Array [
                    Object {
                      "id": "1",
                      "name": "openshift",
                      "product_name": "OpenShift Container Platform",
                      "schema": Object {
                        "fields": Array [
                          Object {
                            "component": "text-field",
                            "name": "field",
                          },
                        ],
                        "title": "openshift",
                      },
                    },
                    Object {
                      "id": "2",
                      "name": "amazon",
                      "product_name": "Amazon Web Services",
                      "schema": Object {
                        "fields": Array [
                          Object {
                            "component": "text-field",
                            "name": "field",
                          },
                        ],
                        "title": "AWS",
                      },
                    },
                    Object {
                      "id": "3",
                      "name": "ansible-tower",
                      "product_name": "Ansible Tower",
                      "schema": Object {
                        "fields": Array [
                          Object {
                            "component": "text-field",
                            "name": "field",
                          },
                        ],
                        "title": "AnsibleTower",
                      },
                    },
                  ],
                },
              ],
              "name": "summary",
              "stepKey": "summary",
              "title": "Review source details",
            },
          ],
          "inModal": true,
          "name": "wizard",
          "title": "Add a source",
        },
      ],
    }
  }
  showFormControls={false}
/>
`;
